torch_geometric/data/feature_store.py

```diff
@@ -0,0 +1,227 @@
+        self.backend = backend
+
+    # Core (CRUD) #############################################################
+
+    @abstractmethod
+    def _put_tensor(self, tensor: FeatureTensorType, attr: TensorAttr) -> bool:
+        r"""Implemented by :obj:`FeatureStore` subclasses."""
+        pass
+
+    def put_tensor(self, tensor: FeatureTensorType, attr: TensorAttr) -> bool:
+        r"""Synchronously adds a :obj:`FeatureTensorType` object to the feature
+        store.
+
+        Args:
+            tensor (FeatureTensorType): the features to be added.
+            attr (TensorAttr): any relevant tensor attributes that correspond
+                to the feature tensor. See the :obj:`TensorAttr` documentation
+                for required and optional attributes. It is the job of
+                implementations of a FeatureStore to store this metadata in a
+                meaningful way that allows for tensor retrieval from a
+                :obj:`TensorAttr` object.
+        Returns:
+            bool: whether insertion was successful.
+        """
+        attr = TensorAttr.cast(attr)
+        assert attr.index is not None
+        assert attr.index.size(dim=0) == tensor.size(dim=-1)
+        return self._put_tensor(tensor, attr)
+
+    @abstractmethod
+    def _get_tensor(self, attr: TensorAttr) -> Optional[FeatureTensorType]:
+        r"""Implemented by :obj:`FeatureStore` subclasses."""
+        pass
+
+    def get_tensor(self, attr: TensorAttr) -> Optional[FeatureTensorType]:
+        r"""Synchronously obtains a :obj:`FeatureTensorType` object from the
+        feature store. Feature store implementors guarantee that the call
+        get_tensor(put_tensor(tensor, attr), attr) = tensor.
+
+        Args:
+            attr (TensorAttr): any relevant tensor attributes that correspond
+                to the tensor to obtain. See :obj:`TensorAttr` documentation
+                for required and optional attributes. It is the job of
+                implementations of a FeatureStore to store this metadata in a
+                meaningful way that allows for tensor retrieval from a
+                :obj:`TensorAttr` object.
+        Returns:
+            FeatureTensorType, optional: a tensor of the same type as the
+            index, or None if no tensor was found.
+        """
+        def to_type(tensor):
+            if tensor is None:
+                return None
+            if isinstance(attr.index, torch.Tensor):
+                return torch.from_numpy(tensor) if isinstance(
+                    tensor, np.ndarray) else tensor
+            if isinstance(attr.index, np.ndarray):
+                return tensor.numpy() if isinstance(tensor,
+                                                    torch.Tensor) else tensor
+            raise ValueError
+
+        attr = TensorAttr.cast(attr)
+        assert attr.index is not None
+
+        return to_type(self._get_tensor(attr))
```

Comment:
It looks like we have to handle the `index` and `tensor` dtype conversion in `_get_tensor` as well. Is this intended?

Reply:
I don't think so, why would we need to do this conversion in `_get_tensor`?

Reply:
Yeah, it might not be possible to do that in `_get_tensor` because we do not know the `dtype` of the tensor in advance. Ultimately, this means that the implementor needs to take care of type conversion as well (to not index a PyTorch tensor with a numpy array and vice versa), so I am not sure I see the point of `to_type` after `_get_tensor` has done its job.

Reply:
Yeah, I wasn't sure about the best way to handle this - one way to handle this would be to always type convert to numpy before passing to `_get_tensor`, but this felt like unnecessary overhead. But the only other option would be to push down this logic entirely to the feature store, which also doesn't seem ideal. Any thoughts?